### PR TITLE
Use TraceIdentity as key for state stores

### PR DESCRIPTION
**Pull request by surajpuvvada**

### CODE REVIEW

**Review by tim-mwangi — DISMISSED** [review]

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java [hunk 1]
```diff
@@ -74,35 +74,50 @@ public void punctuate(long timestamp) {
...
+            .buildStructuredTraceFromRawSpans(rawSpanList, traceId, customerId,
+                timestamps);
+
+        logger.debug("Emit customer_id=[{}], trace_id=[{}], spans_count=[{}]",
+            customerId,
+            HexUtils.getHex(traceId),
+            rawSpanList.size());
+
+        context.forward(null, trace, outputTopicProducer);
+      }
```
Comment (kotharironak) [inline raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java:121]: can this be true `agg.value().getRawSpans().isEmpty()`? I think this should not happen. we can put a log message in else block to check if you want.

**Review by kotharironak — DISMISSED** [review]
lgtm, there is one logging comment in case if you decided to address.

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java [hunk 2]
```diff
@@ -65,7 +77,9 @@ public void init(ProcessorContext context) {
     long currentTimeMs = System.currentTimeMillis();
 
     if (logger.isDebugEnabled()) {
-      logger.debug("Updating ts=[{}] for trace_id=[{}]", Instant.ofEpochMilli(currentTimeMs), key);
+      logger.debug("Updating ts=[{}] for customer_id=[{}], trace_id=[{}]",
```
Comment (kotharironak) [inline raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java]:
```suggestion
      logger.debug("Updating ts=[{}] for tenant_id=[{}], trace_id=[{}]",
```

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java [hunk 3]
```diff
@@ -84,8 +98,10 @@ public void init(ProcessorContext context) {
 
     long traceEmitTs = currentTimeMs + groupingWindowTimeoutMs;
     if (logger.isDebugEnabled()) {
-      logger.debug("Updating trigger_ts=[{}] for trace_id=[{}]", Instant.ofEpochMilli(traceEmitTs),
-          key);
+      logger.debug("Updating trigger_ts=[{}] for for customer_id=[{}], trace_id=[{}]",
```
Comment (kotharironak) [inline raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java]:
```suggestion
      logger.debug("Updating trigger_ts=[{}] for for tenant_id=[{}], trace_id=[{}]",
```

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java [hunk 4]
```diff
@@ -74,35 +81,53 @@ public void punctuate(long timestamp) {
...
         customerId = firstSpan.getCustomerId();
+
+        List<RawSpan> rawSpanList = agg.value().getRawSpans();
+        recordSpansPerTrace(rawSpanList.size(), List.of(Tag.of("tenant_id", customerId)));
+        Timestamps timestamps = trackEndToEndLatencyTimestamps(timestamp, agg);
+        StructuredTrace trace = StructuredTraceBuilder
+            .buildStructuredTraceFromRawSpans(rawSpanList, traceId, customerId,
+                timestamps);
+
+        logger.debug("Emit customer_id=[{}], trace_id=[{}], spans_count=[{}]",
```
Comment (kotharironak) [inline raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java]:
```suggestion
        logger.debug("Emit tenant_id=[{}], trace_id=[{}], spans_count=[{}]",
```

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java [hunk 5]
```diff
@@ -74,35 +81,53 @@ public void punctuate(long timestamp) {
...
+            customerId,
```
Comment (kotharironak) [inline raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java]:
```suggestion
            tenantId,
```

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java [hunk 6]
```diff
@@ -74,35 +81,53 @@ public void punctuate(long timestamp) {
...
-              timestamps);
-
-      logger.debug("Emit trace_id=[{}], spans_count=[{}]", key, rawSpanList.size());
-      context.forward(null, trace, outputTopicProducer);
     } else {
       // implies spans for the trace have arrived within the last 'sessionTimeoutMs' interval
       // so the session inactivity window is extended from the last timestamp
       if (logger.isDebugEnabled()) {
-        logger.debug("Re-scheduling emit trigger for trace_id=[{}] to [{}]", key,
+        logger.debug("Re-scheduling emit trigger for customer_id=[{}], trace_id=[{}] to [{}]",
```
Comment (kotharironak) [inline raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java]:
```suggestion
        logger.debug("Re-scheduling emit trigger for tenant_id=[{}], trace_id=[{}] to [{}]",
```

**Review by kotharironak — COMMENTED** [review]
overall looks fine (if we can rename customerId -> tenantId)

File: raw-spans-grouper/build.gradle.kts [hunk 7]
```diff
@@ -34,11 +34,12 @@ tasks.test {
 }
 
 dependencies {
+    implementation("org.hypertrace.core.spannormalizer:span-normalizer-api:0.1.23-SNAPSHOT")
```
Comment (ravisingal) [inline raw-spans-grouper/build.gradle.kts]:
```suggestion
    implementation("org.hypertrace.core.spannormalizer:span-normalizer-api:0.1.23")
```

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java (hunk 1 above)
Comment (surajpuvvada) [reply raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java:121]: should never happen ...

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java (hunk 2 above)
Comment (surajpuvvada) [reply raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java]: done

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java (hunk 3 above)
Comment (surajpuvvada) [reply raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java]: done

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java (hunk 5 above)
Comment (surajpuvvada) [reply raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java]: done

File: raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java (hunk 6 above)
Comment (surajpuvvada) [reply raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java]: done

File: raw-spans-grouper/src/main/resources/configs/common/application.conf [hunk 8]
```diff
@@ -19,6 +19,10 @@ kafka.streams.config = {
   replication.factor = ${?REPLICATION_FACTOR}
   topic.cleanup.policy = "delete,compact"
 
+  producer.max.request.size = 10485760
+  default.production.exception.handler = org.hypertrace.core.kafkastreams.framework.exceptionhandlers.IgnoreProductionExceptionHandler
```
Comment (kotharironak) [inline raw-spans-grouper/src/main/resources/configs/common/application.conf:23]: If it only requires in production (both production properties) and not as a default, we should move this to helm override.

**Review by kotharironak — APPROVED** [review]

**Comment by tim-mwangi** [comment]
lgtm but will let someone else have the last word.

**Review by buchi-busireddy — APPROVED** [review]

**Pull request merged**